tutorials/turbo_1.ipynb

```diff
@@ -0,0 +1,592 @@
+    "    tr_ub = torch.clamp(x_center + weights * state[\"length\"] / 2.0, 0.0, 1.0)\n",
+    "\n",
+    "    if acqf == \"ts\":\n",
+    "        dim = X.shape[-1]\n",
+    "        sobol = SobolEngine(dim, scramble=True)\n",
+    "        pert = sobol.draw(n_candidates).to(dtype=dtype, device=device)\n",
+    "        pert = tr_lb + (tr_ub - tr_lb) * pert\n",
+    "\n",
+    "        # Create a perturbation mask\n",
+    "        prob_perturb = min(20.0 / dim, 1.0)\n",
+    "        mask = (\n",
+    "            torch.rand(n_candidates, dim, dtype=dtype, device=device)\n",
+    "            <= prob_perturb\n",
+    "        )\n",
+    "        ind = torch.where(mask.sum(dim=1) == 0)[0]\n",
+    "        mask[ind, torch.randint(0, dim - 1, size=(len(ind),), device=device)] = 1\n",
+    "\n",
+    "        # Create candidate points from the perturbations and the mask        \n",
+    "        X_cand = x_center.expand(n_candidates, dim).clone()\n",
+    "        X_cand[mask] = pert[mask]\n",
+    "\n",
+    "        # Sample on the candidate points\n",
+    "        with torch.no_grad(), gpytorch.settings.cholesky_jitter(1e-4):\n",
+    "            y_cand = model(X_cand).sample(torch.Size([batch_size])).t()\n",
+    "\n",
+    "        # Select the best candidates\n",
+    "        X_next = torch.zeros(batch_size, dim, dtype=dtype, device=device)\n",
+    "        for i in range(batch_size):\n",
+    "            indbest = torch.argmax(y_cand[:, i])\n",
+    "            X_next[i, :] = X_cand[indbest, :].clone()\n",
+    "            y_cand[indbest, :] = -float(\"inf\")  # Make sure we never pick this point again\n",
+    "\n",
+    "    elif acqf == \"ei\":\n",
+    "        ei = qExpectedImprovement(model, train_Y.max(), maximize=True)\n",
+    "        X_next, acq_value = optimize_acqf(\n",
+    "            ei,\n",
+    "            bounds=torch.stack([tr_lb, tr_ub]),\n",
+    "            q=batch_size,\n",
+    "            num_restarts=num_restarts,\n",
+    "            raw_samples=raw_samples,\n",
+    "        )\n",
+    "\n",
+    "    return X_next"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Optimization loop\n",
+    "This simple loop runs one instance of TuRBO-1 with Thompson sampling until convergence.\n",
+    "\n",
+    "TuRBO-1 is a local optimizer that can be used for a fixed evaluation budget in a multi-start fashion.  Once TuRBO converges, `state[\"restart_triggered\"]` will be set to true and the run should be aborted.  If you want to run more evaluations with TuRBO, you simply generate a new set of initial points and then keep generating batches until convergence or when the evaluation budget has been exceeded.  It's important to note that evaluations from previous instances are discarded when TuRBO restarts.\n",
+    "\n",
+    "NOTE: We use a `FixedNoiseGP` in this case as the problem is noise-free."
```

Comment:
I am not sure if we want to encourage users to input noisy observations when no observation noise is present.  Inferring the noise can help with numerical stability and help protect against model misspecification (see e.g., https://arxiv.org/pdf/1007.4580.pdf)

Reply:
The thought here is that the `FixedNoiseGP` with a little bit of noise acts as regularization, but it makes sense to learn. I can switch over to the `SingleTaskGP` and add a constraint + prior to keep the learned noise from being large since we know the problem is noise-free.